src/domains/shielder/stores/getShielderIndexedDB.ts

```diff
@@ -3,162 +3,337 @@ import { openDB, IDBPDatabase } from 'idb';
+
+  const handleDatabaseError = async (error: unknown): Promise<IDBPDatabase<DBSchema>> => {
+    const errorMessage = (error as Error).message || '';
+
+    // Check if it's an object store not found error
+    if (errorMessage.includes('object stores was not found') ||
+      errorMessage.includes('object store') ||
+      (error as Error).name === 'NotFoundError') {
+      console.warn('IndexedDB object stores not found. Recreating database...');
+      await deleteDatabase();
+      cache.clear();
+      return await initDB();
+    }
+
+    throw error;
+  };
+
+  const initDB = async (): Promise<IDBPDatabase<DBSchema>> => {
+    const cacheKey = 'db';
+    const cached = cache.get(cacheKey);
+    if (cached) {
+      return cached;
+    }
+
+    const dbPromise = (async () => {
+      try {
+        return await createDB();
+      } catch (error) {
+        // Remove from cache on error so next call will retry
+        cache.delete(cacheKey);
+
+        if ((error as Error).name === 'VersionError') {
+          console.warn('IndexedDB version mismatch. Clearing database and retrying...');
+          await deleteDatabase();
+          return await createDB();
+        }
+        throw error;
+      }
+    })();
+
+    cache.set(cacheKey, dbPromise);
+
+    try {
+      return await dbPromise;
+    } catch (error) {
+      // Remove from cache if promise rejects
+      cache.delete(cacheKey);
+      throw error;
+    }
+  };
+
+  const resetCache = () => {
+    cache.clear();
+  };
+
+  return { initDB, handleDatabaseError, resetCache };
+};
+
+const dbManager = createDBManager();
+const initDB = dbManager.initDB;
+const handleDatabaseError = dbManager.handleDatabaseError;
+const resetDBInstance = dbManager.resetCache;
+
 export const getShielderIndexedDB = (chainId: number, privateKey: Address) => {
   const chainKey = chainId.toString();
-  const hashedPrivateKey = sha256(privateKey);
-  const initDbPromise = initDB();
+  const hashedKey = sha256(privateKey);
+  const dbp = initDB();
 
   return {
     getItem: async (itemKey: string): Promise<string | null> => {
-      const db = await initDbPromise;
-      const allDataForAddress = await db.get(STORE_CLIENTS, hashedPrivateKey);
-      const chainData = allDataForAddress?.[chainKey];
-      return chainData?.[itemKey] ?? null;
+      try {
+        const db = await dbp;
+        const addrData = await db.get(STORE_CLIENTS, hashedKey);
+        return addrData?.[chainKey]?.[itemKey] ?? null;
+      } catch (error) {
+        console.error('Failed to get item from IndexedDB:', error);
+        try {
+          const db = await handleDatabaseError(error);
+          const addrData = await db.get(STORE_CLIENTS, hashedKey);
+          return addrData?.[chainKey]?.[itemKey] ?? null;
+        } catch (retryError) {
+          console.error('Failed to get item from IndexedDB after retry:', retryError);
+          return null;
+        }
+      }
```

Comment:
There is a lot of duplication here and in all `getItem` and `setItem` methods. Why have you decided it's all needed? Could retries be handled on client side?

IIUC, it's critical for shielder tx to be saved. Is this why we have those fallbacks?

In case of multiple simultaneous request are they synchronized? It's seems to me might be deleting and recreating dbs before requests can finish, which in turn will trigger more recreating :sweat_smile: 

Do you have an understanding of what will happen in case of failure during e.g. shielder sync? Is it the case we want to handle here or all we want to do is to upgrade the db safely? In latter case I think the error handling got a bit out of hand :sweat_smile: